### DISCUSSION TITLE
Neper installation on HPC

### DISCUSSION THREAD

**Discussion by xshang93** [neperfepx/neper#345] [Q&A] [answered]
Hi, I'm trying to install neper on to the HPC of our university. When it comes to "make install", I got the following error:

```
Install the project...
-- Install configuration: "Release"
-- Installing: /home/<path_name>/neper/src/bin/neper
CMake Error at cmake_install.cmake:45 (file):
  file RPATH_CHANGE could not write new RPATH:

  to the file:
 /home/<path_name>/neper/src/bin/neper

  Error opening file for update.
```
I checked the neper file being generated in /src/bin/, and it's read and executable only. The neper manual says to install Neper as root, but I'm not a SU of our HPC. Please note that `/home/<path_name>/neper/src/bin/` is my local path.

Is there a way to solve this?

Thank you.

**Comment by xshang93** [comment, marked as answer]
Ok. So I have found a solution, and I thought I'd share it here in case anyone in the future also looking to have Neper installed on a cluster.
The error is caused by RPATH_CHANGE trying to change the original RPATH to a new one, which is an empty one. When I looked at cmake_install.cmake in `neper/src/build/`, I found the following:
```
  if(EXISTS "$ENV{DESTDIR}${CMAKE_INSTALL_PREFIX}/bin/neper" AND
     NOT IS_SYMLINK "$ENV{DESTDIR}${CMAKE_INSTALL_PREFIX}/bin/neper")
    file(RPATH_CHANGE
         FILE "$ENV{DESTDIR}${CMAKE_INSTALL_PREFIX}/bin/neper"
         OLD_RPATH "<home path>/neper/src/build/contrib/scotch:<home path>/neper/src/build/contrib/nlopt:"
         NEW_RPATH "")
```
My understanding is that cmake is trying to change the RPATH to scotch and nlopt (I was using the built-in versions) to an empty directory. Judging from the `if` statement, this happens because I changed my installation directory to my own project folder, which should be a must for most working with a cluster, as you don't have access to other shared areas in the file system.

The solution is simple enough - just comment out where RPATH is being changed, like so:
```
  if(EXISTS "$ENV{DESTDIR}${CMAKE_INSTALL_PREFIX}/bin/neper" AND
     NOT IS_SYMLINK "$ENV{DESTDIR}${CMAKE_INSTALL_PREFIX}/bin/neper")
#    file(RPATH_CHANGE
#         FILE "$ENV{DESTDIR}${CMAKE_INSTALL_PREFIX}/bin/neper"
#         OLD_RPATH "<home path>/neper/src/build/contrib/scotch:<home path>/neper/src/build/contrib/nlopt:"
#        NEW_RPATH "")
```

Et voilà!

Disclaimer: I'm not by any means someone who knows about software compiling. All these are concluded from my search & experimenting. There might be obvious mistakes but it worked for me, and I certainly hope it'll help others in the same situation.

Cheers,
XS

**Reply by rquey** [reply]
Thanks for the update.

On a cluster (or other machines) with no root access, instead of running `make install`, I would just copy the `neper` binary to a `bin` directory located in my home directory (and have it included in my `PATH`)...

**Reply by xshang93** [reply]
Didn't think that is a thing! If it can be done this simple, what's the advantage of compiling it?

**Reply by rquey** [reply]
`make` compiles, and `make install` copies the resulting binary to a system directory. I was talking about replacing `make install` here. You still need to compile, of course.

**Reply by xshang93** [reply]
Right. As you can see, I'm showing my ignorance on this part. So my question rephrases as, "if it can be done this simple, why are we instructed to `make install` rather than just copy the binary?"

**Reply by xshang93** [reply]
by the way, there is another issue I raised that is both Neper and FEPX related. I posted it in FEPX discussion but it would be great if you could take a look. https://github.com/acmelab-ua/FEPX/discussions/59

Thank you, as always, for your help!

**Reply by rquey** [reply]
> Right. As you can see, I'm showing my ignorance on this part. So my question rephrases as, "if it can be done this simple, why are we instructed to `make install` rather than just copy the binary?"

The common practice is to copy the binary to a system directory to make it available system-wide. Copying to a local folder makes it available to you only, and the local folder must be included in your `PATH`...
